File: backend/controllers/ticketController.js
const Ticket = require('../models/Ticket');
const { sendTicketStatusUpdateEmail , sendTicketAssignedEmail} = require('../services/emailService');
const User = require('../models/User');
// Créer un ticket
const createTicket = async (req, res) => {
  const { title, description } = req.body;

  try {
    const ticket = await Ticket.create({
      title,
      description,
      createdBy: req.user.id // 🔥 On prend l'utilisateur connecté grâce à protect
    });
     // 2. Récupérer l'email depuis la base de données via l'ID
     const user = await User.findById(req.user.id);
     ticket.status = 'open';

     // 3. Vérifier que l'utilisateur existe bien et a un email
     if (user && user.email) {
       console.log(`📧 Email pour notification création ticket: ${user.email}`);
       await sendTicketStatusUpdateEmail(user.email, ticket._id, ticket.status);
     } else {
       console.warn('⚠️ Utilisateur introuvable ou sans email');
     }

    res.status(201).json(ticket);
  } catch (error) {
     next(error);
  }  
};

// Récupérer tous les tickets créés par l'utilisateur connecté
const getMyTickets = async (req, res) => {
  try {
    const tickets = await Ticket.find({ createdBy: req.user.id })
    .populate('assignedTo', 'name email role');
    
    res.status(200).json(tickets);
  } catch (error) {
    next(error);
  }  
};
// Récupérer tous les tickets assignés à l'utilisateur connecté
const getAssignedTickets = async (req, res) => {
  try {
    const tickets = await Ticket.find({ assignedTo: req.user.id })
      .populate('assignedTo', 'name email role')   // Pour afficher qui est assigné
      .populate('createdBy', 'name email role');   // Optionnel : pour afficher qui a créé le ticket

    res.status(200).json(tickets);
  } catch (error) {
    next(error);
  }
};

// Assigner un ticket à un agent
const assignTicket = async (req, res) => {
    const { id } = req.params; // ID du ticket
    const { agentId } = req.body; // ID de l'agent qu'on veut assigner
  
    try {
      const ticket = await Ticket.findById(id);
      if (!ticket) {
        const error = new Error('Ticket not found');
        error.statusCode = 404;
        return next(error);

      }
  
      ticket.assignedTo = agentId;
      ticket.status = 'assigned'; // Optionnel : on change aussi le statut
      await ticket.save();
      const creator = await User.findById(ticket.createdBy);
      if (creator && creator.email) {
        await sendTicketStatusUpdateEmail(creator.email, ticket._id, ticket.status);
      } else {
        console.error('Créateur non trouvé ou sans email');
      }
      // 2. Récupérer l'email depuis la base de données via l'ID
     const Agent = await User.findById(agentId);

     const sender = await User.findById(req.user.id); // 👈 Ajout ici
     const senderName = sender.name;
     const senderRole = sender.role;

     if (Agent && Agent.email) {
      console.log(`📧 Email d'agent pour notification création ticket: ${Agent.email}`);
      await sendTicketAssignedEmail(Agent.email, ticket._id,senderName, senderRole );
    } else {
      console.warn('⚠️ agent introuvable ou sans email');
    }
     

      res.status(200).json(ticket);
    } catch (error) {
      next(error);
    }
  };
  const updateTicket = async (req, res) => {
    const { id } = req.params;
    const { title, description } = req.body;
  
    try {
      const ticket = await Ticket.findById(id);
      if (!ticket) {
        return res.status(404).json({ message: 'Ticket not found' });
      }
  
      const isCreator = ticket.createdBy.toString() === req.user.id;
      const isAssignedToMe = ticket.assignedTo?.toString() === req.user.id;
      const isAdmin = req.user.role === 'admin';
  
      // 🌐 Logique de droits :
      if (isAdmin) {
        // ✅ admin peut tout faire
      } else if (req.user.role === 'user') {
        // ❌ si le ticket est fermé ou assigné, l'user ne peut plus le modifier
        if (!isCreator || ticket.status === 'closed' || ticket.assignedTo) {
          return res.status(403).json({ message: 'Accès refusé : vous ne pouvez plus modifier ce ticket.' });
        }
      } else if (req.user.role === 'agent') {
        if (!isAssignedToMe) {
          return res.status(403).json({ message: 'Accès refusé : vous n\'êtes pas assigné à ce ticket.' });
        }
      } // ✅ FIN DU else if
       
      // ✏️ Appliquer les modifications
      if (title) ticket.title = title;
      if (description) ticket.description = description;
  
      const creator = await User.findById(ticket.createdBy);
      if (creator?.email) {
        await sendTicketStatusUpdateEmail(creator.email, ticket._id, ticket.status);
      }
  
      await ticket.save();
      res.status(200).json(ticket);
  
    } catch (error) {
      console.error("Erreur updateTicket:", error);
      res.status(500).json({ message: 'Erreur serveur' });
    }
  };  
  // supprimer un ticket par admin 
  const deleteTicket = async (req, res) => {
    const { id } = req.params;
  
    try {
      const ticket = await Ticket.findById(id);
      if (!ticket) {
        return res.status(404).json({ message: 'Ticket not found' });
      }
  
      // Seul un administrateur peut supprimer un ticket
      if (req.user.role !== 'admin') {
        return res.status(403).json({ message: 'Seul un administrateur peut supprimer un ticket.' });
      }
  
      await ticket.deleteOne(); // ou ticket.remove() si tu utilises Mongoose < 6
      res.status(200).json({ message: 'Ticket supprimé avec succès.' });
    } catch (error) {
      console.error("Erreur deleteTicket:", error);
      res.status(500).json({ message: 'Erreur serveur lors de la suppression.' });
    }
  };
  // Clôturer un ticket
const closeTicket = async (req, res) => {
    const { id } = req.params;
  
    try {
      const ticket = await Ticket.findById(id);
  
      if (!ticket) {
        const error = new Error('Ticket not found');
        error.statusCode = 404;
        return next(error);
      }
  
      // Vérifier si c'est le créateur ou l'agent assigné qui clôture sauf admin
      if (
        //ticket.createdBy.toString() !== req.user.id &&
        ticket.assignedTo?.toString() !== req.user.id &&
        req.user.role !== 'admin'
      ) {
        const error = new Error('Not authorized to close this ticket');
        error.statusCode = 403;
        return next(error);
      }
  
      ticket.status = 'closed';
      await ticket.save();
      const creator = await User.findById(ticket.createdBy);

    if (creator && creator.email) {
      await sendTicketStatusUpdateEmail(creator.email, ticket._id, ticket.status);
    } else {
      console.error('Créateur non trouvé ou sans email');
    }

      res.status(200).json(ticket);
    } catch (error) {
      next(error);
    }
  };
  const getAllTickets = async (req, res) => {
    try {
      const tickets = await Ticket.find().populate('createdBy assignedTo', 'name email role');
      res.status(200).json(tickets);
    } catch (error) {
      next(error);
    }
  };
  const getTicketById = async (req, res, next) => {
    const { id } = req.params;
  
    try {
      const ticket = await Ticket.findById(id).populate('createdBy assignedTo', 'name email role');
  
      if (!ticket) {
        const error = new Error('Ticket not found');
        error.statusCode = 404;
        return next(error);
      }
  
      res.status(200).json(ticket);
    } catch (error) {
      next(error);
    }
  };
  
  module.exports = {
    createTicket,
    getMyTickets,
    getAssignedTickets,
    assignTicket,
    updateTicket,
    closeTicket,
    getAllTickets,
    getTicketById,
    deleteTicket

  };
  

File: backend-dot.net/Controllers/TicketController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using SupportTicketApp.Data;
using SupportTicketApp.Models;
using SupportTicketApp.Services.Email;
using System.Security.Claims;
using SupportTicketApp.DTOs;


namespace SupportTicketApp.Controllers;

[ApiController]
[Route("api/tickets")]
[Authorize]
public class TicketController : ControllerBase
{
    private readonly ApplicationDbContext _context;
    private readonly IEmailService _email;

    public TicketController(ApplicationDbContext context, IEmailService email)
    {
        _context = context;
        _email = email;
    }

    private int GetUserId() => int.Parse(User.FindFirstValue(ClaimTypes.NameIdentifier));
    private string GetUserRole() => User.FindFirstValue(ClaimTypes.Role);

[HttpPost]
[Authorize(Roles = "user")]
public async Task<IActionResult> CreateTicket([FromBody] TicketDto dto)
{
    var userId = GetUserId();

    var ticket = new Ticket
    {
        Title = dto.Title,
        Description = dto.Description,
        Status = "open",
        CreatedAt = DateTime.Now,
        UpdatedAt = DateTime.Now,
        CreatedById = userId
    };

    _context.Tickets.Add(ticket);
    await _context.SaveChangesAsync();

    var creator = await _context.Users.FindAsync(userId);
    if (creator?.Email != null)
        await _email.SendTicketStatusUpdateEmail(creator.Email, ticket.Id, ticket.Status);

    return CreatedAtAction(nameof(GetTicketById), new { id = ticket.Id }, ticket);
}

    [HttpGet("my")]
    public async Task<IActionResult> GetMyTickets()
    {
        var id = GetUserId();
        var tickets = await _context.Tickets
            .Where(t => t.CreatedById == id)
            .Include(t => t.AssignedTo)
            .ToListAsync();

        return Ok(tickets);
    }

    [HttpGet("assigned")]
    public async Task<IActionResult> GetAssignedTickets()
    {
        var id = GetUserId();
        var tickets = await _context.Tickets
            .Where(t => t.AssignedToId == id)
            .Include(t => t.CreatedBy)
            .Include(t => t.AssignedTo)
            .ToListAsync();

        return Ok(tickets);
    }

    [HttpGet("all")]
    [Authorize(Roles = "admin,agent")]
    public async Task<IActionResult> GetAllTickets()
    {
        var tickets = await _context.Tickets
            .Include(t => t.CreatedBy)
            .Include(t => t.AssignedTo)
            .ToListAsync();

        return Ok(tickets);
    }

    [HttpGet("{id}")]
    public async Task<IActionResult> GetTicketById(int id)
    {
        var ticket = await _context.Tickets
            .Include(t => t.CreatedBy)
            .Include(t => t.AssignedTo)
            .FirstOrDefaultAsync(t => t.Id == id);

        if (ticket == null) return NotFound(new { message = "Ticket not found" });
        return Ok(ticket);
    }

    [HttpPut("{id}/assign")]
    [Authorize(Roles = "admin")]
    public async Task<IActionResult> AssignTicket(int id, [FromBody] AssignDto request)
    {
        var ticket = await _context.Tickets.FindAsync(id);
        if (ticket == null) return NotFound(new { message = "Ticket not found" });

        // 🚫 Bloquer l'assignation si le ticket est fermé
        if (ticket.Status == "closed")
        return BadRequest(new { message = "Cannot assign a closed ticket." });

        ticket.AssignedToId = request.AgentId;
        ticket.Status = "assigned";
        await _context.SaveChangesAsync();

        var creator = await _context.Users.FindAsync(ticket.CreatedById);
        if (creator?.Email != null)
            await _email.SendTicketStatusUpdateEmail(creator.Email, ticket.Id, ticket.Status);

        var agent = await _context.Users.FindAsync(request.AgentId);
        var sender = await _context.Users.FindAsync(GetUserId());
        if (agent?.Email != null)
            await _email.SendTicketAssignedEmail(agent.Email, ticket.Id, sender.Name, sender.Role);

        return Ok(ticket);
    }

    [HttpPut("{id}")]
    public async Task<IActionResult> UpdateTicket(int id, [FromBody] TicketUpdateDto updated)
    {
        var ticket = await _context.Tickets.FindAsync(id);
        if (ticket == null) return NotFound(new { message = "Ticket not found" });

        var userId = GetUserId();
        var role = GetUserRole();

        var isCreator = ticket.CreatedById == userId;
        var isAssigned = ticket.AssignedToId == userId;

        if (role == "admin" || 
            (role == "user" && isCreator && ticket.Status != "closed" && ticket.AssignedToId == null) || 
            (role == "agent" && isAssigned))
        {
            ticket.Title = updated.Title ?? ticket.Title;
            ticket.Description = updated.Description ?? ticket.Description;
            ticket.UpdatedAt = DateTime.Now;

            await _context.SaveChangesAsync();

            var creator = await _context.Users.FindAsync(ticket.CreatedById);
            if (creator?.Email != null)
                await _email.SendTicketStatusUpdateEmail(creator.Email, ticket.Id, ticket.Status);

            return Ok(ticket);
        }

        return Forbid();
    }

    [HttpPut("{id}/close")]
    public async Task<IActionResult> CloseTicket(int id)
    {
        var ticket = await _context.Tickets.FindAsync(id);
        if (ticket == null) return NotFound(new { message = "Ticket not found" });

        var role = GetUserRole();
        var userId = GetUserId();

        if (role == "admin" || ticket.AssignedToId == userId)
        {
            ticket.Status = "closed";
            await _context.SaveChangesAsync();

            var creator = await _context.Users.FindAsync(ticket.CreatedById);
            if (creator?.Email != null)
                await _email.SendTicketStatusUpdateEmail(creator.Email, ticket.Id, ticket.Status);

            return Ok(ticket);
        }

        return Forbid();
    }

    [HttpDelete("{id}")]
    [Authorize(Roles = "admin")]
    public async Task<IActionResult> DeleteTicket(int id)
    {
        var ticket = await _context.Tickets.FindAsync(id);
        if (ticket == null) return NotFound(new { message = "Ticket not found" });

        _context.Tickets.Remove(ticket);
        await _context.SaveChangesAsync();

        return Ok(new { message = "Ticket supprimé avec succès." });
    }

    
}
